core/state/access_witness.go

```diff
@@ -20,391 +20,235 @@ import (
-func (aw *AccessWitness) touchAddress(addr []byte, isWrite bool) (bool, bool, bool, bool, bool) {
-	var (
-		stem                                [31]byte
-		stemRead, selectorRead              bool
-		stemWrite, selectorWrite, chunkFill bool
-	)
-	copy(stem[:], addr[:31])
-
-	// Check for the presence of the stem
-	if _, hasStem := aw.Branches[stem]; !hasStem {
-		stemRead = true
-		aw.Branches[stem] = AccessWitnessReadFlag
-	}
-
-	// Check for the presence of the leaf selector
-	if _, hasSelector := aw.Chunks[common.BytesToHash(addr)]; !hasSelector {
-		selectorRead = true
-		aw.Chunks[common.BytesToHash(addr)] = AccessWitnessReadFlag
-	}
-
-	if isWrite {
-		stemWrite, selectorWrite, chunkFill = aw.touchAddressOnWrite(addr)
-	}
-
-	return stemRead, selectorRead, stemWrite, selectorWrite, chunkFill
-}
-
-func (aw *AccessWitness) touchAddressAndChargeGas(addr []byte, isWrite bool) uint64 {
-	var gas uint64
-
-	stemRead, selectorRead, stemWrite, selectorWrite, selectorFill := aw.touchAddress(addr, isWrite)
-
-	if stemRead {
-		gas += params.WitnessBranchReadCost
-	}
-	if selectorRead {
-		gas += params.WitnessChunkReadCost
-	}
-	if stemWrite {
-		gas += params.WitnessBranchWriteCost
-	}
-	if selectorWrite {
-		gas += params.WitnessChunkWriteCost
-	}
-	if selectorFill {
-		gas += params.WitnessChunkFillCost
-	}
-
-	return gas
-}
-
-func (aw *AccessWitness) TouchAddressOnWriteAndComputeGas(addr []byte) uint64 {
-	return aw.touchAddressAndChargeGas(addr, true)
-}
-
-func (aw *AccessWitness) TouchAddressOnReadAndComputeGas(addr []byte) uint64 {
-	return aw.touchAddressAndChargeGas(addr, false)
-}
-
 // Merge is used to merge the witness that got generated during the execution
 // of a tx, with the accumulation of witnesses that were generated during the
 // execution of all the txs preceding this one in a given block.
 func (aw *AccessWitness) Merge(other *AccessWitness) {
-	for k := range other.Branches {
-		if _, ok := aw.Branches[k]; !ok {
-			aw.Branches[k] = other.Branches[k]
-		}
-	}
-
-	for k, chunk := range other.Chunks {
-		if _, ok := aw.Chunks[k]; !ok {
-			aw.Chunks[k] = chunk
-		}
+	for k := range other.branches {
+		aw.branches[k] |= other.branches[k]
 	}
-
-	for k, v := range other.InitialValue {
-		if _, ok := aw.InitialValue[k]; !ok {
-			aw.InitialValue[k] = v
-		}
+	for k, chunk := range other.chunks {
+		aw.chunks[k] |= chunk
 	}
-
-	// TODO see if merging improves performance
-	//for k, v := range other.addrToPoint {
-	//if _, ok := aw.addrToPoint[k]; !ok {
-	//aw.addrToPoint[k] = v
-	//}
-	//}
 }
 
 // Key returns, predictably, the list of keys that were touched during the
 // buildup of the access witness.
 func (aw *AccessWitness) Keys() [][]byte {
-	keys := make([][]byte, 0, len(aw.Chunks))
-	for key := range aw.Chunks {
-		var k [32]byte
-		copy(k[:], key[:])
-		keys = append(keys, k[:])
+	// TODO: consider if parallelizing this is worth it, probably depending on len(aw.chunks).
```

Comment:
I left a TODO here just for consideration; I didn't see this appearing in CPU profiles because I think this API is mostly used for proof generation. In any case, just left a reminder idea.

Reply:
Calculating tree-keys ends up calling `PointToHash`, which under hood calls `Bytes()` to hash the point. It doesn't use the usual Point->Fr hashing as we use in go-verkle (that `toFr`). [See here](https://github.com/gballet/go-ethereum/blob/pbss/replay-fixes/trie/utils/verkle.go#L216). 
 
So despite `toFrMultiple` doesn't make sense here, we could still do inverse batching but for other reasons.

I want to see what is slow from here guided by benchmarks. Probably when we include proofs in blocks in the benchmark.